src/OpenTelemetry/Metrics/AggregatorStore.cs

```diff
@@ -370,13 +562,339 @@ private int LookupAggregatorStore(KeyValuePair<string, object?>[] tagKeysAndValu
+        // If self-claimed, then add a fresh entry to the dictionary
+        // If an available MetricPoint is found, then only increment the ReferenceCount
+
+        Debug.Assert(this.tagsToMetricPointIndexDictionaryDelta != null, "this.tagsToMetricPointIndexDictionaryDelta was null");
+
+        // Delete the entry for these Tags and get another MetricPoint.
+        lock (this.tagsToMetricPointIndexDictionaryDelta!)
+        {
+            LookupData? dictionaryValue;
+            if (lookupData.SortedTags != Tags.EmptyTags)
+            {
+                // Check if no other thread added a new entry for the same Tags.
+                // If no, then remove the existing entries.
+                if (this.tagsToMetricPointIndexDictionaryDelta.TryGetValue(lookupData.SortedTags, out dictionaryValue))
+                {
+                    if (dictionaryValue == lookupData)
+                    {
+                        // No other thread added a new entry for the same Tags.
+                        this.tagsToMetricPointIndexDictionaryDelta.TryRemove(lookupData.SortedTags, out _);
+                        this.tagsToMetricPointIndexDictionaryDelta.TryRemove(lookupData.GivenTags, out _);
+                    }
+                    else
+                    {
+                        // Some other thread added a new entry for these Tags. Use the new MetricPoint
+                        lookupData = dictionaryValue;
+                        foundMetricPoint = true;
+                    }
+                }
+            }
+            else
+            {
+                if (this.tagsToMetricPointIndexDictionaryDelta.TryGetValue(lookupData.GivenTags, out dictionaryValue))
+                {
+                    if (dictionaryValue == lookupData)
+                    {
+                        // No other thread added a new entry for the same Tags.
+                        this.tagsToMetricPointIndexDictionaryDelta.TryRemove(lookupData.GivenTags, out _);
+                    }
+                    else
+                    {
+                        // Some other thread added a new entry for these Tags. Use the new MetricPoint
+                        lookupData = dictionaryValue;
+                        foundMetricPoint = true;
+                    }
+                }
+            }
+
+            if (!foundMetricPoint
+                && this.TryGetAvailableMetricPointRare(inputTags, sortedTags, length, out var tempLookupData))
+            {
+                foundMetricPoint = true;
+                lookupData = tempLookupData;
+            }
+        }
+
+        if (foundMetricPoint)
+        {
+            var index = lookupData.Index;
+            ref var metricPointAtIndex = ref this.metricPoints[index];
+            _ = Interlocked.Increment(ref metricPointAtIndex.ReferenceCount);
+            return index;
+        }
```

Comment:
I think in the unusual circumstance of an very short collect cycle it's possible that the metric point acquired under lock here has been released yet again by the snapshot thread prior to incrementing the reference count.

Reply:
I wonder if ReferenceCount should be initialized to 1 instead of 0 when constructed?

Reply:
@alanwest Thanks! That's a good catch!

I think it would be simpler to just put that code inside the lock. That should fix it.

Reply:
I'm not sure putting it inside the lock will fix it. Snapshot does not synchronize on the lock. It just does

```chsharp
if (metricPoint.LookupData != null && Interlocked.CompareExchange(ref metricPoint.ReferenceCount, int.MinValue, 0) == 0)
```

Reply:
Fixed this in https://github.com/open-telemetry/opentelemetry-dotnet/pull/4486/commits/628174d69454cc8c4ebcb92680996750c42e2d86